common/src/value/string.rs

```diff
@@ -0,0 +1,150 @@
+use gc::{Finalize, Gc, Trace};
+use std::borrow::Cow;
+
+#[derive(Trace, Finalize)]
+pub enum JsString {
+    Latin1(Box<[u8]>),
+    Utf16(Box<[u16]>),
+}
+impl JsString {
+    pub fn from_latin1(src: &[u8]) -> Gc<JsString> {
+        let mut copy = Vec::with_capacity(src.len());
+        copy.extend_from_slice(src);
+        Self::from_latin1_owned(copy)
+    }
+    pub fn from_utf8(src: &str) -> Gc<JsString> {
+        if src.is_ascii() {
+            Self::from_latin1(src.as_bytes())
+        } else {
+            let copy = src.encode_utf16().collect();
+            Self::from_utf16_owned(copy)
+        }
+    }
+    pub fn from_utf16(src: &[u16]) -> Gc<JsString> {
+        let mut copy = Vec::with_capacity(src.len());
+        copy.extend_from_slice(src);
+        Self::from_utf16_owned(copy)
+    }
+    pub fn from_latin1_owned(src: Vec<u8>) -> Gc<JsString> {
+        Gc::new(JsString::Latin1(src.into_boxed_slice()))
+    }
+    pub fn from_utf16_owned(src: Vec<u16>) -> Gc<JsString> {
+        Gc::new(JsString::Utf16(src.into_boxed_slice()))
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Latin1(str) => str.len(),
+            Self::Utf16(str) => str.len(),
+        }
+    }
+
+    pub fn as_latin1(&self) -> Result<Cow<[u8]>, ()> {
+        match self {
+            Self::Latin1(str) => Ok(Cow::Borrowed(str)),
+            Self::Utf16(str) => {
+                let mut owned = Vec::with_capacity(str.len());
```

Comment:
It's already using a Vec and iterator, so using Iterator::map and collect might read better (- it should be preallocated and not effect performance).

Stylistic preference, your call.

Reply:
It must return an error if any character is greater than 255, so you can't just use map and collect. We could use `try_collect`, but it's feature-flagged and it's something we can live without.

Reply:
> It must return an error if any character is greater than 255, so you can't just use map and collect. We could use `try_collect`, but it's feature-flagged and it's something we can live without.

`collect::<Result<_>>` is an option.
<https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.collect>
`try_collect` is vaguely different.